### PR TITLE
MEP-0001: Future-based shuffle on Ray

### PR DESCRIPTION
**Pull request by chaokunyang**
This PR proposes enhancement to support Future-based shuffle on Ray, which is the first MEP in our shuffle optimizations:
1. **Future-based shuffle on Ray**
2. Push-based Shuffle using Ray Future
3. Mitigate Straggler tasks
4. Dynamically coalescing shuffle partitions
5. Dynamically splitting shuffle partitions

### CODE REVIEW

File: 2022-05-24 Future-based shuffle for Ray.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+- Introduce a new group of `FetchShuffle` operands which only records `n_mappers` and `n_reducers` to reduce meta overhead:
+```python
+class FetchShuffleType(Enum):
+    FETCH_SHUFFLE_BY_INDEX = 0
+    FETCH_SHUFFLE_BY_DATA_KEY = 1
+
+class FetchShuffleByIndex(Operand):
+    _op_type_ = opcodes.FETCH_SHUFFLE_BY_INDEX
+    n_mappers = Int32Field("n_mappers")
+    n_reducers = Int32Field("n_reducers")
```
Comment (wjsi) [inline 2022-05-24 Future-based shuffle for Ray.md]: Better to clarify how `n_mappers` and `n_reducers` locates input mappers and output reducers. BTW, `n_xxx` refers to number of certain objects, thus it is better to give a new name.

File: 2022-05-24 Future-based shuffle for Ray.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+- Make shuffle operands mapper produce output with deterministic number and order.
+    - `TensorReshape` and `TensorBinCount` mapper needs change. 
+- _bagging.py uses `source_idxes`, which alsoe needs to be refactored.
+
+#### Graph Changes
+Add an option to replace `ShuffleProxy` to `FetchShuffleByIndex` when building subtask graph
+
+#### Using a ShuffleManager to manage Ray shuffle execution
+```python
+class ShuffleManager:
```
Comment (wjsi) [inline 2022-05-24 Future-based shuffle for Ray.md]: Where and how is this manager created, and by whom? Is this manager only applicable on Ray backend and locates inside `mars.services.task`? Please clarify.

File: 2022-05-24 Future-based shuffle for Ray.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+- Setup the foundation for push-based shuffle using ray future
+
+## Design and Architecture
+When the Mars subtask has upstream and downstream dependencies, the downstream subtask's dependence on the output of the upstream subtask is determined by the data key. The input data required for subtask execution is a context dict object from data key to data. 
+
+For non-shuffle operators, the data key is the chunk key; but for shuffle operators, the data key is a tuple of (chunk key, reducer index). 
+
+Therefore, in order to build a Shuffle DAG based on Ray future, either all the data keys output by each mapper subtask are known in advance, or the data keys are removed.
+
+### Why not using data key?
```
Comment (wjsi) [inline 2022-05-24 Future-based shuffle for Ray.md]: Traditionally, `xEP`s add a separate section called `alternative` at the end of the doc to describe abandoned designs. Better to follow it.

File: 2022-05-24 Future-based shuffle for Ray.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    def get_reducer_input_refs(self, subtask) -> List["ray.ObjectRef"]:
+        pass
+
+    def get_n_reducers(self, subtask):
+        pass
+```
+
+#### Build Shuffle DAG code
+Load reducer mapper inputs:
```
Comment (wjsi) [inline 2022-05-24 Future-based shuffle for Ray.md]: More details needed showing how this implementation can coexist with Mars implementation.

**Review by wjsi — CHANGES_REQUESTED** [review]

File: 2022-05-24 Future-based shuffle for Ray.md (hunk 1 above)
Comment (chaokunyang) [reply 2022-05-24 Future-based shuffle for Ray.md]: I think `n_mappers` and `n_reducers` are OK. There are similar namign such as `n_workers`/`n_threads`

File: 2022-05-24 Future-based shuffle for Ray.md (hunk 2 above)
Comment (chaokunyang) [reply 2022-05-24 Future-based shuffle for Ray.md]: added some details about shuffle manager. FYI, it's only used by ray executor and located at mars/services/task/execution/ray/shuffle.py`

File: 2022-05-24 Future-based shuffle for Ray.md (hunk 3 above)
Comment (chaokunyang) [reply 2022-05-24 Future-based shuffle for Ray.md]: Good suggestion, Moved to `alternative` section

File: 2022-05-24 Future-based shuffle for Ray.md (hunk 4 above)
Comment (chaokunyang) [reply 2022-05-24 Future-based shuffle for Ray.md]: The `ShuffleManager` won't be shared by mars, but all operands change can be shared by all backends

**Comment by wjsi** [comment]
Please rename the file as `mep-0001-ray-future-based-shuffle.md`.

File: 2022-05-24 Future-based shuffle for Ray.md (hunk 1 above)
Comment (wjsi) [reply 2022-05-24 Future-based shuffle for Ray.md]: What's the key of this operand and related chunk? Some clarification needed.

Comment (chaokunyang) [reply 2022-05-24 Future-based shuffle for Ray.md]: It's similar to `FetchShuffle` operands, the only difference is that it didn't include `source_keys/source_idxes/source_mappers` info and added `n_mappers `/`n_reducers `

**Comment by fyrestone** [comment]
Some spelling suggestions:
- Unify the `ray` and `Ray` to `Ray`.
- Use `C++` instead of `c++`.

Besides, can you add some comments to the `ShuffleManager` API?

File: mep-0001-ray-future-based-shuffle.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+- Improve scheduling performance by using ray c++ scheduler
+- Support shuffle failover by lineage recontruction
+- Setup the foundation for push-based shuffle using ray future
+
+## Design and Architecture
+When the Mars subtask has upstream and downstream dependencies, the downstream subtask's dependence on the output of the upstream subtask is determined by the data key. The input data required for subtask execution is a context dict object from data key to data. 
+
+For non-shuffle operators, the data key is the chunk key; but for shuffle operators, the data key is a tuple of (chunk key, reducer index). 
+
+Therefore, in order to build a Shuffle DAG based on Ray future, either all the data keys output by each mapper subtask are known in advance, or the data keys are removed.
```
Comment (fyrestone) [inline mep-0001-ray-future-based-shuffle.md]:
```suggestion
Therefore, in order to build a Shuffle DAG based on Ray future, either all the data keys output by each mapper subtask are known in advance, or the data keys can be calculated from the subtask (chunk graph).
```

File: 2022-05-24 Future-based shuffle for Ray.md (hunk 1 above)
Comment (wjsi) [reply 2022-05-24 Future-based shuffle for Ray.md]: For me, it is more convenient to include a key of `ShuffleProxy` operand to reference the whole shuffle group.

File: mep-0001-ray-future-based-shuffle.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+### Building shuffle DAG by reducer index
+Acutally, Mars' ChunkGraph has information about the upstream and downstream chunk dependencies. After converting to Ray DAG, this part of the information still exists. The reason for relying on data keys is that Mars replaced `ShuffleProxy` with `FetchShuffle`. The implementation of `Fetch/FetchShuffle` is an abstraction of P2P, which is actually inconsistent with the semantics of Ray DAG. Therefore, when using Ray Backend, a new set of Shuffle operators can be introduced to remove FetchShuffle and rely on Ray to automatically resolve subtask input arguments.
```
Comment (fyrestone) [inline mep-0001-ray-future-based-shuffle.md]:
```suggestion
Acutally, Mars' ChunkGraph has information about the upstream and downstream chunk dependencies. After converting to Ray DAG, this part of the information still exists. The reason for relying on data keys is that Mars replaced `ShuffleProxy` with `FetchShuffle`. The implementation of `Fetch` or `FetchShuffle` is an abstraction of P2P, which is actually inconsistent with the semantics of Ray DAG. Therefore, when using Ray Backend, a new set of Shuffle operators can be introduced to replace the `FetchShuffle` and rely on Ray to automatically resolve subtask dependencies.
```

File: mep-0001-ray-future-based-shuffle.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    source_keys = ListField("source_keys", FieldTypes.string)
+    source_idxes = ListField("source_idxes", FieldTypes.tuple(FieldTypes.uint64))
+    source_mappers = ListField("source_mappers", FieldTypes.uint16)
+```
+- Make shuffle operands mapper produce output with deterministic number and order.
+    - `TensorReshape` and `TensorBinCount` mapper needs change. 
+- _bagging.py uses `source_idxes`, which alsoe needs to be refactored.
+
+#### Graph Changes
+Add an option to replace `ShuffleProxy` to `FetchShuffleByIndex` when building subtask graph
```
Comment (fyrestone) [inline mep-0001-ray-future-based-shuffle.md]:
```suggestion
Add an option to replace `ShuffleProxy` to `FetchShuffleByIndex` when building subtask graph.
```

File: mep-0001-ray-future-based-shuffle.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+#### Using a ShuffleManager to manage Ray shuffle execution
+`ShuffleManager` is defined in `mars/services/task/execution/ray/shuffle.py` and only used by ray executor. When `TaskProcessorActor` invokes `RayTaskExecutor.execute_subtask_graph` to execute a subtask graph, `RayTaskExecutor` will create `ShuffleManager` to manage shuffle execution.
```
Comment (fyrestone) [inline mep-0001-ray-future-based-shuffle.md]:
```suggestion
`ShuffleManager` is defined in `mars/services/task/execution/ray/shuffle.py` and only used by ray executor. When `TaskProcessorActor` invokes `RayTaskExecutor.execute_subtask_graph` to execute a subtask graph, the `RayTaskExecutor` will create a `ShuffleManager` to manage shuffle execution.
```

File: mep-0001-ray-future-based-shuffle.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+`ShuffleManager` will do following things:
+- Build mapper index which is a dict from mapper subtask to (`shuffle_index`, `mapper_index`). A subtask graoh may have multiple groups of shuffle, `shuffle_index` indicates a mapper subtask belongs to which shuffle. `mapper_index` indicates is oridinal in all mapper subtasks of current shuffle.
+- Build reducer index which is a dict from reducer subtask to (`shuffle_index`, `reducer_ordinal`). A subtask graoh may have multiple groups of shuffle, `shuffle_index` indicates a reducer subtask belongs to which shuffle. `reducer_ordinal` indicates is oridinal in all reducer subtasks of current shuffle. If some reducers missing, `ShuffleManager` will filling None for those reducers index.
```
Comment (fyrestone) [inline mep-0001-ray-future-based-shuffle.md]:
```suggestion
- Build reducer index which is a dict from reducer subtask to (`shuffle_index`, `reducer_ordinal`). A subtask graoh may have multiple groups of shuffle, `shuffle_index` indicates a reducer subtask belongs to which shuffle. `reducer_ordinal` indicates is oridinal in all reducer subtasks of current shuffle. If some reducers are missing, `ShuffleManager` will fill `None` for those reducers index.
```

File: mep-0001-ray-future-based-shuffle.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+- Recording mapper output object refs for all mapper subtasks, which will be used by later reducers get all mapper inputs based on reducer oridnal.
```
Comment (fyrestone) [inline mep-0001-ray-future-based-shuffle.md]:
```suggestion
- Recording mapper output object refs for all mapper subtasks, which will be used by later reducers to get all the mapper inputs based on reducer oridnal.
```

Comment (chaokunyang) [reply mep-0001-ray-future-based-shuffle.md]: `all mapper inputs` is more idiomatic

**Comment by chaokunyang** [comment]
@fyrestone @wjsi @qinxuye @zhongchun @hekaisheng @yuyiming  I've updated the proposal, could you please review it again?

File: 2022-05-24 Future-based shuffle for Ray.md (hunk 1 above)
Comment (chaokunyang) [reply 2022-05-24 Future-based shuffle for Ray.md]: You mean set the key of `ShuffleProxy` to `FetchShuffleByIndex` operand? Sounds acceptable to me.

File: mep-0001-ray-future-based-shuffle.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
+# Motivation
+Shuffle is the core operation for large-scale data processing. The data is partitioned according to the partition key in each record through the ALL-TO-ALL connection, and sent from all mappers to all reducers. Due to the nature of its ALL-TO_ALL connections, shuffle often becomes the bottleneck of the performance, stability and scalability of the entire system. This MEP introduces a pull-based shuffle implementation using Ray future which serves following goals:
```
Comment (zhongchun) [inline mep-0001-ray-future-based-shuffle.md]:
```suggestion
Shuffle is the core operation for large-scale data processing. The data is partitioned according to the partition key in each record through the ALL-TO-ALL connection, and sent from all mappers to all reducers. Due to the nature of its ALL-TO-ALL connections, shuffle often becomes the bottleneck of the performance, stability and scalability of the entire system. This MEP introduces a pull-based shuffle implementation using Ray future which serves following goals:
```

File: mep-0001-ray-future-based-shuffle.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+   # number of all upstream mappers
+   n_mappers = Int32Field("n_mappers")
+   # number of all downstream reducers
+   n_reducers = Int32Field("n_reducers")
+ 
+class FetchShuffle(Operand):
+   _op_type_ = opcodes.FETCH_SHUFFLE_BY_DATA_KEY
+   source_keys = ListField("source_keys", FieldTypes.string)
+   source_idxes = ListField("source_idxes", FieldTypes.tuple(FieldTypes.uint64))
+   source_mappers = ListField("source_mappers", FieldTypes.uint16)
```
Comment (zhongchun) [inline mep-0001-ray-future-based-shuffle.md:26]: Use `FetchShuffleByIndex` to replace `FetchShuffle`? If so, `FetchShuffle` can be removed.

Comment (chaokunyang) [reply mep-0001-ray-future-based-shuffle.md:26]: `FetchShuffle` will still be used by mars shuffle

**Review by zhongchun — DISMISSED** [review]
LGTM.

File: mep-0001-ray-future-based-shuffle.md (hunk 12 above)
Comment (qinxuye) [reply mep-0001-ray-future-based-shuffle.md:26]: Can we just record the n_mappers and n_reducers in `FetchShuffle`? and I think for Mars shuffle we can remove the `source_keys` stuff as well, but for compatibility we can mix them together, I think we don't need a new opcode.

Comment (chaokunyang) [reply mep-0001-ray-future-based-shuffle.md:26]: If mars can remove `source_keys ` stuff too, add `n_mappers` and `n_reducers` would be better.

**Review by qinxuye — DISMISSED** [review]
LGTM

**Review by yuyiming — APPROVED** [review]
LGTM

**Review by fyrestone — APPROVED** [review]
LGTM

File: mep-0001-ray-future-based-shuffle.md [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+
+        Parameters
+        ----------
+        subtask
+        output_object_refs : List["ray.ObjectRef"]
+            Mapper output ObjectRefs.
+        """
+        pass
+
+    def get_reducer_input_refs(self, subtask: Subtask) -> List["ray.ObjectRef"]:
```
Comment (fyrestone) [inline mep-0001-ray-future-based-shuffle.md:86]: Maybe we can move the `get_reducer_input_refs` and `get_n_reducers` to the `SubtaskGraph` in the future, then Ray backend can share the same logic with Mars backend.

Comment (chaokunyang) [reply mep-0001-ray-future-based-shuffle.md:86]: Actually, those refs are Ray ObjectRefs, and in future there will be more ray-specific Ray Object management  code when push-based shuffle is implemented for ray. I think it's not feasible to share this logic with Mars backend

Comment (fyrestone) [reply mep-0001-ray-future-based-shuffle.md:86]: I know the refs are Ray object refs. For Mars, they are the map data keys. Mars can use this logic to calculate the map data keys.

Comment (chaokunyang) [reply mep-0001-ray-future-based-shuffle.md:86]: When push-based shuffle are implemented, parts of those refs are mapper ouputs, parts of it are merge outputs. I think those logic shouldn't be put into SubtaskGraph

**Review by hekaisheng — APPROVED** [review]
LGTM

**Review by wjsi — APPROVED** [review]
LGTM

**Pull request merged**